### PR TITLE
Implement MSC3966: Add a push rule condition to search for a value in an array.

### PR DESCRIPTION
**Pull request by clokep**
[MSC3966](https://github.com/matrix-org/matrix-spec-proposals/pull/3966) defines an `exact_event_property_contains` condition which allows searching the arrays in events for exact matches.

### CODE REVIEW

File: rust/src/push/mod.rs [hunk 1]
```diff
@@ -280,6 +280,37 @@ impl<'source> FromPyObject<'source> for SimpleJsonValue {
...
+    Value(SimpleJsonValue),
+}
+
+impl<'source> FromPyObject<'source> for JsonValue {
+    fn extract(ob: &'source PyAny) -> PyResult<Self> {
+        if let Ok(l) = <PyList as pyo3::PyTryFrom>::try_from(ob) {
+            if let Ok(a) = l.iter().map(SimpleJsonValue::extract).collect() {
+                Ok(JsonValue::Array(a))
+            } else {
+                // TOOD How to get a good error?
```
Comment (MatMaul) [inline rust/src/push/mod.rs:297]: Should we remove the TODO ? if not, typo :)

**Review by MatMaul — APPROVED** [review]
Minor comment, LGTM.

File: rust/src/push/mod.rs (hunk 1 above)
Comment (DMRobertson) [reply rust/src/push/mod.rs:297]: I'm guessing (without looking closely) that `.collect()` gives you a `Result`. If so, the `if let Ok(a)` discards the Error part of that result before you can do anything with it.

I think you want to either `match` on that Result, or use one of the fancy methods on Result (.map, .map_err?).

Comment (clokep) [reply rust/src/push/mod.rs]: Thanks @DMRobertson! Switching to `match` made it doable. Thought I had tried that but who knows...

**Pull request merged**